micro_dl/torch_unet/networks/Unet25D.py

```diff
@@ -0,0 +1,309 @@
+        1.) https://elifesciences.org/articles/55502
+
+        Architecture takes in stack of 2d inputs given as a 3d tensor and returns a 2d interpretation.
+        Learns 3d information based upon input stack, but speeds up training by compressing 3d
+        information before the decoding path. Uses interruption conv layers in the Unet skip paths to
+        compress information with z-channel convolution.
+
+        :param int in_channels: number of feature channels in (1 or more)
+        :param int out_channels: number of feature channels out (1 or more)
+        :param int input_stack_depth: depth of input stack in z
+        :param int output_stack_depth: depth of output stack
+        :param int/tuple(int, int) xy_kernel_size: size of x and y dimensions of conv kernels in blocks
+        :param bool residual: see name
+        :param float dropout: probability of dropout, between 0 and 0.5
+        :param int num_blocks: number of convolutional blocks on encoder and decoder paths
+        :param int num_block_layers: number of layer sequences repeated per block
+        :param list[int] num_filters: list of filters/feature levels at each conv block depth
+        :param str task: network task (for virtual staining this is regression): 'seg','reg'
+        :param str debug_mode: if true logs features at each step of architecture, must be manually set
+        """
+        super(Unet25d, self).__init__()
+        self.in_channels = in_channels
+        self.num_blocks = num_blocks
+        self.kernel_size = xy_kernel_size
+        self.residual = residual
+        assert (
+            dropout >= 0 and dropout <= 0.5
+        ), f"Dropout {dropout} not in allowed range: [0, 0.5]"
+        self.dropout = dropout
+        self.task = task
+        self.debug_mode = False
+
+        # ----- set static parameters ----- #
+        self.block_padding = "same"
+        down_mode = "avgpool"  # TODO set static avgpool
+        up_mode = "trilinear"  # TODO set static trilinear
+        activation = "relu"  # TODO set static relu
+        self.bottom_block_spatial = False  # TODO set static
+
+        # ----- Standardize Filter Sequence ----- #
+        if len(num_filters) != 0:
+            assert len(num_filters) == num_blocks, (
+                "Length of num_filters must be equal to num_"
+                "blocks + 1 (number of convolutional blocks per path)."
+            )
+            self.num_filters = num_filters
+        else:
+            self.num_filters = [pow(2, i) * 16 for i in range(num_blocks + 1)]
+            self.num_filters
+        forward_filters = [in_channels] + self.num_filters
+        backward_filters = [
+            self.num_filters[-(i + 1)] + self.num_filters[-(i + 2)]
+            for i in range(len(self.num_filters))
+            if i < len(self.num_filters) - 1
+        ] + [out_channels]
+
+        # ----- Downsampling steps -----#
+        self.down_list = []
+        if down_mode == "maxpool":
```

Comment:
Instead of a number of if statements with for loops embedded in them, it would be cleaner to map valid strings to nn functions. If you'd like to make it easy for yourself you can require your down_mode to match ff function names (e.g. instead of 'maxpool' -> 'MaxPool3d'), in which case you can use getattr directly:
pooling = getattr(nn, down_mode)
for i in range(num_blocks):
    self.down_list.append(pooling(kernel_size=(1, 2, 2)))

That way you can let the user directly choose from torch.nn which type to use. Although you might want to check first that it's within a set that works with the model so things don't break if the user tries to call a 1D pooling or something.

If you would like to keep down_modes like 'maxpool' you can define a dictionary mapping, e.g.:
pool_dict = {'maxpool': 'MaxPool3d', 'avgpool': 'AvgPool3d'}
and then proceed with getattr.